### PR TITLE
Hts 402

### PR DESCRIPTION
**Pull request by Jackie1988**

### CODE REVIEW

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.helptosavefrontend.connectors
+
+import com.google.inject.{ImplementedBy, Inject, Singleton}
+import org.joda.time.Period
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: do we need to use joda ?  better use java.time if we can

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+@ImplementedBy(classOf[EmailVerificationConnectorImpl])
+trait EmailVerificationConnector {
+
+  def verifyEmail(nino: String, newEmail: String)(implicit hc: HeaderCarrier): Future[VerifyEmailStatus]
+  def isVerified(email: String)(implicit hc: HeaderCarrier): Future[Either[VerifyEmailStatus, Boolean]]
+}
+
+@Singleton
+class EmailVerificationConnectorImpl @Inject() (http: WSHttp, conf: Configuration) extends EmailVerificationConnector with ServicesConfig with Logging {
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: you dont need to inject WSHttp -  use `val http: WSHttpExtension = WSHttp`

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+  override def verifyEmail(nino: String, newEmail: String)(implicit hc: HeaderCarrier): Future[VerifyEmailStatus] = {
+    conf.getInt("services.email-verification.linkTTLMinutes") match {
+      case Some(linkTTLMinutes) ⇒ verify(nino, newEmail, Period.minutes(linkTTLMinutes))
+      case None ⇒ throw new RuntimeException("email-verification.link.ttlMinutes not present in configuration")
+    }
+  }
+
+  private def verify(nino: String, newEmail: String, ttlPeriod: Period)(implicit hc: HeaderCarrier): Future[VerifyEmailStatus] = {
+    val postURL = s"""$verifyEmailURL"""
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: you can simply say `val postURL = verifyEmailURL`

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    val json: JsValue = Json.toJson(EmailVerificationRequest(newEmail, nino, "", ttlPeriod.toString, ttlPeriod.toString, Map()))
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: i think we dont need to convert case class to json , you can directly pass object to POST , look at other connectorrs

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+          RequestNotValidError(nino)
+        case CONFLICT ⇒
+          logger.info("[EmailVerification] - Email already verified")
+          AlreadyVerified(nino, newEmail)
+        case SERVICE_UNAVAILABLE ⇒
+          logger.warn("[EmailVerification] - Email Verification service not available")
+          VerificationServiceUnavailable()
+        case _ ⇒
+          logger.warn("[EmailVerification] - Unexpected status received from email verification")
+          BackendError("Unexpected response from email verification service. Status = " + response.status + ", body = " + response.body.toString)
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: for string concatination, use $  instead of + ?

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    val postURL = s"""$verifyEmailURL"""
+    val json: JsValue = Json.toJson(EmailVerificationRequest(newEmail, nino, "", ttlPeriod.toString, ttlPeriod.toString, Map()))
+
+    http.POST(postURL, json).map { (response: HttpResponse) ⇒
+      response.status match {
+        case OK | CREATED =>
+          logger.info("[EmailVerification] - Successful return of data")
+          Verifing(nino, newEmail)
+        case BAD_REQUEST ⇒
+          logger.warn("[EmailVerification] - Bad Request")
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: you can say `Bad Request for email verification`

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+  }
+
+  private def verify(nino: String, newEmail: String, ttlPeriod: Period)(implicit hc: HeaderCarrier): Future[VerifyEmailStatus] = {
+    val postURL = s"""$verifyEmailURL"""
+    val json: JsValue = Json.toJson(EmailVerificationRequest(newEmail, nino, "", ttlPeriod.toString, ttlPeriod.toString, Map()))
+
+    http.POST(postURL, json).map { (response: HttpResponse) ⇒
+      response.status match {
+        case OK | CREATED =>
+          logger.info("[EmailVerification] - Successful return of data")
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: `successfully triggered to email verification` ..... etc

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+@Singleton
+class EmailVerificationConnectorImpl @Inject() (http: WSHttp, conf: Configuration) extends EmailVerificationConnector with ServicesConfig with Logging {
+
+  val emailVerifyBaseURL = baseUrl("email-verification")
+  val verifyEmailURL = s"$emailVerifyBaseURL/email-verification/verification-requests"
+  def isVerifiedURL(email: String) = s"$emailVerifyBaseURL/email-verification/verified-email-addresses/$email"
+
+  val continueURL = "email-verification.continue.baseUrl"
+  //+ EmailVerificationController.showSuccess.
+  val defaultTemplate = "hts_email_verification"
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: where are you using the above two vals ?

File: app/uk/gov/hmrc/helptosavefrontend/controllers/EmailVerificationController.scala [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+import play.api.mvc.Results._
+import play.api.mvc.{Action, AnyContent}
+import uk.gov.hmrc.helptosavefrontend.views.html.introduction.about_help_to_save
+import uk.gov.hmrc.play.http.HeaderCarrier
+
+import scala.concurrent.Future
+
+
+trait EmailVerificationController {
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/controllers/EmailVerificationController.scala]: why there is no impl yet for this controller ? 
if you want to implement it later , better remove this from this merge -  or its upto you -- its just my personal preference

File: app/uk/gov/hmrc/helptosavefrontend/controllers/EmailVerificationController.scala [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.helptosavefrontend.controllers
+
+import play.api.mvc.Results._
+import play.api.mvc.{Action, AnyContent}
+import uk.gov.hmrc.helptosavefrontend.views.html.introduction.about_help_to_save
+import uk.gov.hmrc.play.http.HeaderCarrier
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/controllers/EmailVerificationController.scala]: unused imports

File: app/uk/gov/hmrc/helptosavefrontend/models/EmailVerificationRequest.scala [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+                                    linkExpiryDuration: String, continueUrl: String,
+                                    templateParameters: Map[String, String])
+
+object EmailVerificationRequest {
+  implicit val formats: Format[EmailVerificationRequest] = Json.format[EmailVerificationRequest]
+
+  def apply(email: String, nino: String, templateId: String, linkExpiryDuration: String, continueUrl: String,
+            templateParameters: Map[String, String]): EmailVerificationRequest =
+    EmailVerificationRequest(email, templateId, linkExpiryDuration, continueUrl,
+      Map("email" → email, "nino" → nino) ++ templateParameters)
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/models/EmailVerificationRequest.scala]: it doesnt make sense to me to pass `nino` for email verification

File: test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+
+import scala.concurrent.Future
+
+class EmailVerificationConnectorSpec extends UnitSpec with TestSupport
+{
+
+  val nino = "AE123XXXX"
+  val email = "email@gmail.com"
+
+  def postHttpMock(returnedStatus: Int, returnedData: Option[JsValue]): WSHttp = {
```
Comment (sureshhmrc) [inline test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala]: you can rename this to `mockPost` -  just my preference

File: test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+    val mockHttp = mock[WSHttp]
+    (mockHttp.POST(_: String, _: JsValue, _: Seq[(String, String)])(_: Writes[Any], _: HttpReads[Any], _: HeaderCarrier)).expects(*, *, *, *, *, *)
+      .returning(Future.successful(HttpResponse(returnedStatus, returnedData)))
+    mockHttp
+  }
+
+  def getHttpMock(returnedStatus: Int, returnedData: Option[JsValue]): WSHttp = {
```
Comment (sureshhmrc) [inline test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala]: `mockGet`

File: test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+    val mockHttp = mock[WSHttp]
+    (mockHttp.GET(_: String) (_: HttpReads[Any], _: HeaderCarrier)).expects(*, *, *)
+      .returning(Future.successful(HttpResponse(returnedStatus, returnedData)))
+    mockHttp
+  }
+
+  def getConfigurationMock(minutes: Int) = {
```
Comment (sureshhmrc) [inline test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala]: `mockConfig`

File: test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+      val http = postHttpMock(Status.SERVICE_UNAVAILABLE, None)
+      val conf = getConfigurationMock(120)
+      val connector = new EmailVerificationConnectorImpl(http, conf)
+      await(connector.verifyEmail(nino, email)) shouldBe VerifyEmailStatus.VerificationServiceUnavailable()
+    }
+
+    "does Period produce ISO 8601 duration syntax when requested" in {
+      //30 minutes = PT30M
+      val thirtyMins = Period.minutes(30).toString  shouldBe "PT30M"
+    }
```
Comment (sureshhmrc) [inline test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala]: are you unit testing joda lib ?

File: test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+import play.api.libs.json.{JsValue, Writes}
+import uk.gov.hmrc.helptosavefrontend.TestSupport
+import uk.gov.hmrc.helptosavefrontend.models.VerifyEmailStatus
+import uk.gov.hmrc.play.http.ws._
+import uk.gov.hmrc.play.http.{HeaderCarrier, HttpReads, HttpResponse}
+import uk.gov.hmrc.play.test.UnitSpec
+
+import scala.concurrent.Future
+
+class EmailVerificationConnectorSpec extends UnitSpec with TestSupport
```
Comment (sureshhmrc) [inline test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala]: TestSupport already extends UnitSpec

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala (hunk 6 above)
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: you'll never get a this status or a 500 in this block because the `POST` method on `http` will throw - use  `post` from `uk.gov.hmrc.helptosavefrontend.config.WSHttpExtension` instead to be able to handle these

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+@Singleton
+class EmailVerificationConnectorImpl @Inject() (http: WSHttp, conf: Configuration) extends EmailVerificationConnector with ServicesConfig with Logging {
+
+  val emailVerifyBaseURL = baseUrl("email-verification")
+  val verifyEmailURL = s"$emailVerifyBaseURL/email-verification/verification-requests"
+  def isVerifiedURL(email: String) = s"$emailVerifyBaseURL/email-verification/verified-email-addresses/$email"
+
+  val continueURL = "email-verification.continue.baseUrl"
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: feels like this should be read from config

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+  //+ EmailVerificationController.showSuccess.
+  val defaultTemplate = "hts_email_verification"
+
+  override def verifyEmail(nino: String, newEmail: String)(implicit hc: HeaderCarrier): Future[VerifyEmailStatus] = {
+    conf.getInt("services.email-verification.linkTTLMinutes") match {
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: this value from config should be read once outside of this function rather than being read every time the function is called

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala (hunk 3 above)
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: don't need triple quotes here either

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala (hunk 4 above)
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: if the TTL is reasd from config you could just have one val outside of this def which represents the TTL as a String rather than toStringing it every time you call this def

Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: also, you're putting in the TTL as the continue URL

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+  }
+
+  def isVerified(email: String)(implicit hc: HeaderCarrier): Future[Either[VerifyEmailStatus, Boolean]] = {
+    val getURL = isVerifiedURL(email)
+    http.GET(getURL).map { response ⇒
+      response.status match {
+        case OK ⇒
+          logger.info("[EmailVerification] - email is verified")
+          Right(true)
+        case NOT_FOUND ⇒
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: you'll not get these statuses again here - use `get` on the `WSHttpExtension`

File: app/uk/gov/hmrc/helptosavefrontend/controllers/EmailVerificationController.scala [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+
+import scala.concurrent.Future
+
+
+trait EmailVerificationController {
+
+  def showSuccess(): Action[AnyContent]  = Action.async {
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavefrontend/controllers/EmailVerificationController.scala]: doesn't need to be `async` if you're not doing anything in a `Future`

File: app/uk/gov/hmrc/helptosavefrontend/controllers/EmailVerificationController.scala [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+    implicit request =>
+      Future.successful(Ok("hello"))
+  }
+}
+
+object EmailVerificationController extends EmailVerificationController {
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavefrontend/controllers/EmailVerificationController.scala]: this object is empty - it's not needed

File: app/uk/gov/hmrc/helptosavefrontend/models/EmailVerificationRequest.scala (hunk 11 above)
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavefrontend/models/EmailVerificationRequest.scala]: shouldn't the email be encrypted?

File: test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala [hunk 22]
```diff
@@ -0,0 +1,141 @@
...
+{
+
+  val nino = "AE123XXXX"
+  val email = "email@gmail.com"
+
+  def postHttpMock(returnedStatus: Int, returnedData: Option[JsValue]): WSHttp = {
+    val mockHttp = mock[WSHttp]
```
Comment (AndyHWChung) [inline test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala]: don't need to create a mock everytime the function is called - create one once outside of the def

File: test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+    (mockHttp.POST(_: String, _: JsValue, _: Seq[(String, String)])(_: Writes[Any], _: HttpReads[Any], _: HeaderCarrier)).expects(*, *, *, *, *, *)
```
Comment (AndyHWChung) [inline test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala]: should probably check the inputs to the post like the url, the body and the headers

File: test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala [hunk 24]
```diff
@@ -0,0 +1,141 @@
...
+      .returning(Future.successful(HttpResponse(returnedStatus, returnedData)))
+    mockHttp
+  }
+
+  def getHttpMock(returnedStatus: Int, returnedData: Option[JsValue]): WSHttp = {
+    val mockHttp = mock[WSHttp]
+    (mockHttp.GET(_: String) (_: HttpReads[Any], _: HeaderCarrier)).expects(*, *, *)
```
Comment (AndyHWChung) [inline test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala]: again, we should probably be checking the URL

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala (hunk 8 above)
Comment (Jackie1988) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: These are now being used.

File: test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala [hunk 25]
```diff
@@ -0,0 +1,141 @@
...
+    val mockConf = mock[Configuration]
+    (mockConf.getInt(_: String)).expects("services.email-verification.linkTTLMinutes").returning(Some(minutes))
+    mockConf
+  }
+
+  "verifyEmail" should {
+    "does good json equal 201" in {
+      val http = postHttpMock(Status.OK, None)
+      val conf = getConfigurationMock(120)
+      val connector = new EmailVerificationConnectorImpl(http, conf)
```
Comment (AndyHWChung) [inline test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala]: do we need a new connector every test?

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala (hunk 3 above)
Comment (Jackie1988) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: Updated.

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala (hunk 18 above)
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: also, instead of using JODA time and reading minutes you can do this in the config:
```
x = 30 minutes
```
then you can use the configs library to do
```
  import configs.syntax._
  val ttl: FiniteDuration = conf.underlying.get[FiniteDuration]("x").valueOrElse(sys.error("Could not read x"))
```

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala (hunk 4 above)
Comment (Jackie1988) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: Yes, thank you, now updated.

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala (hunk 7 above)
Comment (Jackie1988) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: Updated

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala (hunk 6 above)
Comment (Jackie1988) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: updated

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala [hunk 26]
```diff
@@ -0,0 +1,103 @@
...
+
+  private def verify(nino: String, newEmail: String, ttlPeriod: Period)(implicit hc: HeaderCarrier): Future[VerifyEmailStatus] = {
+    val postURL = s"""$verifyEmailURL"""
+    val json: JsValue = Json.toJson(EmailVerificationRequest(newEmail, nino, "", ttlPeriod.toString, ttlPeriod.toString, Map()))
+
+    http.POST(postURL, json).map { (response: HttpResponse) ⇒
+      response.status match {
+        case OK | CREATED =>
+          logger.info("[EmailVerification] - Successful return of data")
+          Verifing(nino, newEmail)
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: seems like this is the happy case and the rest are error cases - using an `Either` here with this being the `Right` and the rest as `Left`s would make this clearer

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala (hunk 5 above)
Comment (Jackie1988) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: Updated

File: app/uk/gov/hmrc/helptosavefrontend/controllers/EmailVerificationController.scala (hunk 10 above)
Comment (Jackie1988) [inline app/uk/gov/hmrc/helptosavefrontend/controllers/EmailVerificationController.scala]: Removed

File: app/uk/gov/hmrc/helptosavefrontend/controllers/EmailVerificationController.scala (hunk 9 above)
Comment (Jackie1988) [inline app/uk/gov/hmrc/helptosavefrontend/controllers/EmailVerificationController.scala]: Removed

File: app/uk/gov/hmrc/helptosavefrontend/models/EmailVerificationRequest.scala (hunk 11 above)
Comment (Jackie1988) [inline app/uk/gov/hmrc/helptosavefrontend/models/EmailVerificationRequest.scala]: Both nino and email are needed for an edge case in verification. We will talk to Jean Sims regarding including the nino in the content.

File: test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala (hunk 16 above)
Comment (Jackie1988) [inline test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala]: Removed

File: test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala (hunk 12 above)
Comment (Jackie1988) [inline test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala]: Updated

File: test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala (hunk 13 above)
Comment (Jackie1988) [inline test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala]: Updated

File: test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala (hunk 14 above)
Comment (Jackie1988) [inline test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala]: Updated

File: test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala (hunk 15 above)
Comment (Jackie1988) [inline test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala]: We just wanted to know what was coming out

File: app/uk/gov/hmrc/helptosavefrontend/models/EmailVerificationRequest.scala (hunk 11 above)
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavefrontend/models/EmailVerificationRequest.scala]: we'll always be able to get the NINO from logging in though

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+trait EmailVerificationConnector {
+
+  def verifyEmail(nino: String, newEmail: String)(implicit hc: HeaderCarrier): Future[Either[VerifyEmailError, Unit]]
+  def isVerified(email: String)(implicit hc: HeaderCarrier): Future[Either[VerifyEmailError, Boolean]]
+}
+
+@Singleton
+class EmailVerificationConnectorImpl @Inject() (http: WSHttp, conf: Configuration) extends EmailVerificationConnector with ServicesConfig with Logging {
+
+  val linkTTLMinutes = conf.getInt("services.email-verification.linkTTLMinutes") match {
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]:
```
val linkTTLMinutes = conf.underlying.getInt("services.email-verification.linkTTLMinutes")
```
basically does the same thing

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala [hunk 28]
```diff
@@ -0,0 +1,99 @@
...
+    case Some(minutes) => minutes
+    case None => throw new RuntimeException("email-verification.link.ttlMinutes not present in configuration")
+  }
+  val emailVerifyBaseURL = baseUrl("email-verification")
+  val verifyEmailURL = s"$emailVerifyBaseURL/email-verification/verification-requests"
+  def isVerifiedURL(email: String) = s"$emailVerifyBaseURL/email-verification/verified-email-addresses/$email"
+
+  val continueURL = "email-verification.continue.baseUrl"
+  //+ EmailVerificationController.showSuccess.
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: doesn't look like this is needed

File: test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala [hunk 29]
```diff
@@ -0,0 +1,127 @@
...
+
+  def mockGet(returnedStatus: Int, email: String, returnedData: Option[JsValue]): Unit = {
+    val verifyEmailURL = s"$emailVerifyBaseURL/email-verification/verified-email-addresses/$email"
+    (mockHttp.get(_: String)(_: HeaderCarrier)).expects(verifyEmailURL, *)
+      .returning(Future.successful(HttpResponse(returnedStatus, returnedData)))
+  }
+  
+  
+  "verifyEmail" should {
+    "does good json equal 201" in {
```
Comment (AndyHWChung) [inline test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala]: don't understand

File: app/uk/gov/hmrc/helptosavefrontend/config/FrontendWiring.scala [hunk 30]
```diff
@@ -16,14 +16,17 @@
 
 package uk.gov.hmrc.helptosavefrontend.config
 
-import play.api.http.HttpVerbs.{GET => GET_VERB, POST => POST_VERB}
+import com.google.inject.{Inject, Singleton}
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/config/FrontendWiring.scala]: Looks ok to me except that we need to change it several places

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala (hunk 27 above)
Comment (Jackie1988) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: updated

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala (hunk 28 above)
Comment (Jackie1988) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: removed

File: test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala (hunk 29 above)
Comment (Jackie1988) [inline test/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnectorSpec.scala]: Updated with better english :)

File: app/uk/gov/hmrc/helptosavefrontend/models/EmailVerificationRequest.scala (hunk 11 above)
Comment (Jackie1988) [inline app/uk/gov/hmrc/helptosavefrontend/models/EmailVerificationRequest.scala]: for now this is how we are checking the user that logged in, is the user being redirected via the email

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala (hunk 17 above)
Comment (Jackie1988) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: the baseURL is in the config, the continueURL has not yet been defined

File: app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala (hunk 19 above)
Comment (Jackie1988) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/EmailVerificationConnector.scala]: now using the WSHttpExtension

**Comment by AndyHWChung** [comment]
LGTM
